sljit_src/sljitNativeS390X.c

```diff
@@ -0,0 +1,2613 @@
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <sys/auxv.h>
+
+#ifdef __ARCH__
+#define ENABLE_STATIC_FACILITY_DETECTION 1
+#else
+#define ENABLE_STATIC_FACILITY_DETECTION 0
+#endif
+#define ENABLE_DYNAMIC_FACILITY_DETECTION 1
+
+SLJIT_API_FUNC_ATTRIBUTE const char* sljit_get_platform_name(void)
+{
+	return "s390x" SLJIT_CPUINFO;
+}
+
+/* Instructions. */
+typedef sljit_uw sljit_ins;
+
+/* Instruction tags (most significant halfword). */
+const sljit_ins sljit_ins_const = (sljit_ins)(1) << 48;
+
+/* General Purpose Registers [0-15]. */
+typedef sljit_uw sljit_gpr;
+const sljit_gpr r0 = 0; /* 0 in address calculations; reserved */
+const sljit_gpr r1 = 1; /* reserved */
+const sljit_gpr r2 = 2; /* 1st argument */
+const sljit_gpr r3 = 3; /* 2nd argument */
+const sljit_gpr r4 = 4; /* 3rd argument */
+const sljit_gpr r5 = 5; /* 4th argument */
+const sljit_gpr r6 = 6; /* 5th argument; 1st saved register */
+const sljit_gpr r7 = 7;
+const sljit_gpr r8 = 8;
+const sljit_gpr r9 = 9;
+const sljit_gpr r10 = 10;
+const sljit_gpr r11 = 11;
+const sljit_gpr r12 = 12;
+const sljit_gpr r13 = 13;
+const sljit_gpr r14 = 14; /* return address and flag register */
+const sljit_gpr r15 = 15; /* stack pointer */
+
+const sljit_gpr tmp0 = 0; /* r0 */
+const sljit_gpr tmp1 = 1; /* r1 */
+
+/* Link registers. The normal link register is r14, but since
+   we use that for flags we need to use r0 instead to do fast
+   calls so that flags are preserved. */
+const sljit_gpr link_r = 14;     /* r14 */
+const sljit_gpr fast_link_r = 0; /* r0 */
+
+/* Flag register layout:
+
+   0               32  33  34      36      64
+   +---------------+---+---+-------+-------+
+   |      ZERO     | 0 | 0 |  C C  |///////|
+   +---------------+---+---+-------+-------+
+*/
+const sljit_gpr flag_r = 14; /* r14 */
+
+static const sljit_gpr reg_map[SLJIT_NUMBER_OF_REGISTERS + 1] = {
+	2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 15
+};
+
+struct sljit_s390x_const {
+	struct sljit_const const_; /* must be first */
+	sljit_sw init_value;       /* required to build literal pool */
+};
+
+/* Convert SLJIT register to hardware register. */
+static sljit_gpr gpr(sljit_s32 r)
+{
+	SLJIT_ASSERT(r != SLJIT_UNUSED);
+	r -= SLJIT_R0; /* normalize */
+	SLJIT_ASSERT(r < (sljit_s32)(sizeof(reg_map) / sizeof(reg_map[0])));
+	return reg_map[r];
+}
+
+/* Size of instruction in bytes. Tags must already be cleared. */
+static sljit_uw sizeof_ins(sljit_ins ins)
+{
+	if (ins == 0) return 2; /* keep faulting instructions */
+	if ((ins & 0x00000000ffff) == ins) return 2;
+	if ((ins & 0x0000ffffffff) == ins) return 4;
+	if ((ins & 0xffffffffffff) == ins) return 6;
```

Comment:
no complains in s390x (neither gcc or clang), which also support __int128, so guess it still has a long way to go.

I added suffix to some cases that could benefit from it, but left the rest for later as I wanted to avoid unnecessary churn in the first change (which is big enough already)